# Assignment 1 exercise 1
### Yagublu Jeyhun 


In [23]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from sklearn import preprocessing
import seaborn as sns
import matplotlib.pyplot as plt
import math
import time

from scipy.stats.stats import pearsonr
import itertools

from sklearn.decomposition import PCA

pd.set_option('display.expand_frame_repr', False)
%matplotlib inline

In [24]:
data_path = "corrtestdata.csv"
original_data = data = pd.read_csv(data_path)


###### Categorical:
* gender
* femstate
* kmethod
* sulcer
* place

###### Numerical:
  **RatioScale:**
   * liverind
   * heartind
   * appind
   * gonfatind
   * batind
   * tailind
   * ADWBind
   * gonind
   * BMI
   * blength
   * weight
    
 **Interval scale** :
   * year
 
 **Circular Interval** :
   * day

### Exercise 1:
### A)  
We calculate the correlations. I have used the library called pingouin which calculates the Pairwise correlation using Pearson method.

In [25]:
import pingouin as pg
correlation=pg.pairwise_corr(data.drop('id',axis=1), method='pearson')


#### i)
We try to see what is the biggest correlations involving categorical features.
* First we create a new dataframe which consists of correlations where at least one feature is categorical.
* We calculate absolute values of r . Since it is between [-1,1] and absolute value of r determines how much data are correlated ( even if negatively )
* then we sort them in descending order

In [26]:
categorical=['gender','femstate','kmethod','sulcer','place']
pd.set_option('display.max_rows', 500)
correlation_categorical=correlation[(correlation['X'].isin([categorical]) | correlation['Y'].isin(categorical))]
correlation_categorical=correlation_categorical.reset_index(drop=True)


In [27]:

correlation_categorical.sort_values('r',ascending=False,inplace=True)
correlation_categorical=correlation_categorical.reset_index(drop=True)



In [28]:
correlation_categorical.loc[:,'abs_r']=correlation_categorical.loc[:,'r'].apply(abs)

In [29]:

correlation_categorical.sort_values('abs_r',ascending=False,inplace=True)
correlation_categorical=correlation_categorical.reset_index(drop=True)
correlation_categorical 


X         Y   method alternative    n         r           CI95%          p-unc        BF10     power     abs_r
0      gender  femstate  pearson   two-sided  563 -0.889752  [-0.91, -0.87]  3.119115e-193  1.342e+189  1.000000  0.889752
1     kmethod     place  pearson   two-sided  563  0.773357     [0.74, 0.8]  3.975520e-113  2.332e+109  1.000000  0.773357
2         day     place  pearson   two-sided  563  0.703150    [0.66, 0.74]   3.913850e-85   3.273e+81  1.000000  0.703150
3   gonfatind     place  pearson   two-sided  563  0.571116    [0.51, 0.62]   4.749538e-50   4.412e+46  1.000000  0.571116
4   gonfatind   kmethod  pearson   two-sided  563  0.492583    [0.43, 0.55]   9.486289e-36   2.871e+32  1.000000  0.492583
5         day   kmethod  pearson   two-sided  563  0.452667    [0.38, 0.52]   8.623685e-30   3.602e+26  1.000000  0.452667
6    liverind   kmethod  pearson   two-sided  563 -0.322689  [-0.39, -0.25]   4.140844e-15   1.175e+12  1.000000  0.322689
7    liverind     place  pearson   two-sided  563 -0.310915  [-0.38, -0.23]   4.393870e-14   1.158e+11  1.000000  0.310915
8    femstate     place  pearson   two-sided  563  0.300979    [0.22, 0.37]   2.975794e-13   1.776e+10  1.000000  0.300979
9    femstate   kmethod  pearson   two-sided  563  0.262931    [0.18, 0.34]   2.347481e-10   2.623e+07  0.999995  0.262931
10        day  femstate  pearson   two-sided  563  0.262647    [0.18, 0.34]   2.458130e-10   2.508e+07  0.999995  0.262647
11     gender     place  pearson   two-sided  563 -0.250108  [-0.33, -0.17]   1.774838e-09   3.663e+06  0.999979  0.250108
12        day    gender  pearson   two-sided  563 -0.249396  [-0.33, -0.17]   1.979560e-09   3.295e+06  0.999977  0.249396
13        day    sulcer  pearson   two-sided  563 -0.244304  [-0.32, -0.16]   4.277596e-09   1.559e+06  0.999960  0.244304
14     weight     place  pearson   two-sided  563  0.239918    [0.16, 0.32]   8.193437e-09   8.298e+05  0.999937  0.239918
15     appind     place  pearson   two-sided  563  0.221344     [0.14, 0.3]   1.119082e-07   6.614e+04  0.999625  0.221344
16     appind  femstate  pearson   two-sided  563  0.205820    [0.13, 0.28]   8.410770e-07    9482.833  0.998583  0.205820
17     batind   kmethod  pearson   two-sided  563  0.195181    [0.11, 0.27]   3.073512e-06    2737.773  0.996763  0.195181
18     weight   kmethod  pearson   two-sided  563  0.189065    [0.11, 0.27]   6.275449e-06    1384.104  0.994949  0.189065
19     sulcer     place  pearson   two-sided  563 -0.186668  [-0.27, -0.11]   8.250089e-06    1066.208  0.994023  0.186668
20    tailind     place  pearson   two-sided  563  0.181015     [0.1, 0.26]   1.551367e-05     584.422  0.991224  0.181015
21   liverind  femstate  pearson   two-sided  563  0.155040    [0.07, 0.23]   2.217037e-04      47.438  0.959228  0.155040
22     gender   kmethod  pearson   two-sided  563 -0.145773  [-0.23, -0.06]   5.210156e-04      21.375  0.935396  0.145773
23    blength     place  pearson   two-sided  563  0.142234    [0.06, 0.22]   7.128159e-04      15.979  0.923877  0.142234
24     batind     place  pearson   two-sided  563  0.139040    [0.06, 0.22]   9.402366e-04      12.367  0.912203  0.139040
25   heartind    sulcer  pearson   two-sided  563  0.127927    [0.05, 0.21]   2.356704e-03       5.316  0.861364  0.127927
26     appind   kmethod  pearson   two-sided  563  0.118983     [0.04, 0.2]   4.699293e-03        2.84  0.808219  0.118983
27  gonfatind    sulcer  pearson   two-sided  563 -0.107967  [-0.19, -0.03]   1.035947e-02       1.399  0.728084  0.107967
28     appind    sulcer  pearson   two-sided  563 -0.105706  [-0.19, -0.02]   1.208602e-02       1.221  0.709827  0.105706
29     weight    sulcer  pearson   two-sided  563 -0.105518  [-0.19, -0.02]   1.224090e-02       1.207  0.708280  0.105518
30     weight    gender  pearson   two-sided  563 -0.092577  [-0.17, -0.01]   2.805783e-02       0.585  0.594390  0.092577
31   femstate    sulcer  pearson   two-sided  563 -0.064969   [-0.15, 0.02] 

**We can see that**:
*  gender and femstate are strongly negatively correlated r= -0.8898
* kmethod and place are strongly positively correlated  r= 0.773
* place and day are positively correlated 0.7031
* sulcer and day are little negatively correlated r= -0.224

#### ii)
We try to see what is the biggest correlations involving temporal features day and year.
* First we create a new dataframe which consists of correlations where at least one feature is temproal .
* We repeat previous steps for i)


In [30]:
temporal=['day','year']
correlation_temporal=correlation[(correlation['X'].isin(temporal) | correlation['Y'].isin(temporal))]
correlation_temporal=correlation_temporal.reset_index(drop=True)


In [31]:
correlation_temporal.loc[:,'abs_r']=correlation_temporal.loc[:,'r'].apply(abs)
correlation_temporal.sort_values('abs_r',ascending=False,inplace=True)
correlation_temporal=correlation_temporal.reset_index(drop=True)
correlation_temporal.head()

X        Y   method alternative    n         r         CI95%         p-unc       BF10  power     abs_r
0      day    place  pearson   two-sided  563  0.703150  [0.66, 0.74]  3.913850e-85  3.273e+81    1.0  0.703150
1  kmethod     year  pearson   two-sided  563  0.542852   [0.48, 0.6]  1.826650e-44  1.262e+41    1.0  0.542852
2    place     year  pearson   two-sided  563  0.461829  [0.39, 0.52]  4.331073e-31  6.959e+27    1.0  0.461829
3      day  kmethod  pearson   two-sided  563  0.452667  [0.38, 0.52]  8.623685e-30  3.602e+26    1.0  0.452667
4      day   gonind  pearson   two-sided  563  0.441341  [0.37, 0.51]  3.077189e-28  1.048e+25    1.0  0.441341

**We can see that**:
*  day and place are strongly  correlated r= 0.703150
* kmethod and year are  positively correlated  r= 0.542852


#### iii)
We try to see what is the biggest correlations involving other numerical featureas.
* We already have Correlations dataframe so we just add abs_r (absolute value of r) feature and sort to find out the strongest correlations


In [32]:
correlation.loc[:,'abs_r']=correlation.loc[:,'r'].apply(abs)


In [33]:
correlation.sort_values('abs_r',ascending=False,inplace=True)
correlation=correlation.reset_index(drop=True)
correlation.head()


X         Y   method alternative    n         r           CI95%          p-unc        BF10  power     abs_r
0   gender  femstate  pearson   two-sided  563 -0.889752  [-0.91, -0.87]  3.119115e-193  1.342e+189    1.0  0.889752
1   weight   blength  pearson   two-sided  563  0.878782     [0.86, 0.9]  2.217358e-182   2.09e+178    1.0  0.878782
2   weight       BMI  pearson   two-sided  563  0.877609     [0.86, 0.9]  2.775997e-181  1.686e+177    1.0  0.877609
3  kmethod     place  pearson   two-sided  563  0.773357     [0.74, 0.8]  3.975520e-113  2.332e+109    1.0  0.773357
4      day     place  pearson   two-sided  563  0.703150    [0.66, 0.74]   3.913850e-85   3.273e+81    1.0  0.703150

**We can see that**:
*  weight and blength are strongly negatively correlated r= 0.889752
*  BMI and weight are  strongly positively correlated  r= 0.878782
*  gonfatind and gonind are positivvely correlated r=0.679151


### B)
Remove outlier rats, rat2, rat53, rat120, and rat434, and calculate correlations again. How did the correlations involving either
liverind or heartind change? What is your explanation for big
changes?


In [34]:
pd.set_option('display.max_rows', 600)
data_before=data.copy()
data.head()

id  day  weight  gender  liverind  heartind    appind  femstate  gonfatind    batind  sulcer  kmethod   tailind  blength  place  year   ADWBind    gonind       BMI
0  rat1   67    22.0       1  0.036364  0.005091  0.013636         4   0.000000  0.001864       1        2  0.714286     10.5      1     5  0.454545  0.000000  0.199546
1  rat2  251   182.0       1  0.010440  0.004396  0.008791         4   0.023077  0.000742       3        2  0.666667     19.5      3     3  0.302198  1.648659  0.478632
2  rat3  230    37.5       1  0.051200  0.006133  0.020800         4   0.000000  0.001467       2        1  0.904762     10.5      2     8  0.653333  0.000000  0.340136
3  rat4  261    45.0       2  0.047333  0.005556  0.105111        -1   0.293333  0.002178       1        1  0.869565     11.5      4     2  0.260000  2.653242  0.340265
4  rat5  262    43.0       1  0.052558  0.005349  0.021628         4   0.000000  0.001884       1        1  0.720000     12.5      2     8  0.500000  0.000000  0.275200

* First we compute correlations involving liverind and heartind before the removal

In [35]:
correlation=pg.pairwise_corr(data.drop('id',axis=1), method='pearson')
correlation.head()

X         Y   method alternative    n         r           CI95%         p-unc       BF10     power
0  day    weight  pearson   two-sided  563  0.116523     [0.03, 0.2]  5.638231e-03       2.41  0.791683
1  day    gender  pearson   two-sided  563 -0.249396  [-0.33, -0.17]  1.979560e-09  3.295e+06  0.999977
2  day  liverind  pearson   two-sided  563 -0.124649  [-0.21, -0.04]  3.050469e-03      4.202  0.843169
3  day  heartind  pearson   two-sided  563 -0.011001   [-0.09, 0.07]  7.945040e-01      0.055  0.057783
4  day    appind  pearson   two-sided  563  0.222792     [0.14, 0.3]  9.199697e-08  7.991e+04  0.999671

* I create a function which return correlations involving only given feautures

In [36]:
def correlation_given_feautures (correlation , feautures):
    correlation_feautures=correlation[(correlation['X'].isin(feautures) | correlation['Y'].isin(feautures))]
    correlation_feautures=correlation_feautures.reset_index(drop=True)
    correlation_feautures.loc[:,'abs_r']=correlation_feautures.loc[:,'r'].apply(abs)
    correlation_feautures.sort_values('abs_r',ascending=False,inplace=True)
    correlation_feautures=correlation_feautures.reset_index(drop=True)
    return correlation_feautures

In [37]:
livheart_0=['liverind','heartind']
correlation_livheart_0=correlation_given_feautures(correlation, livheart_0)
correlation_livheart_0

X          Y   method alternative    n         r           CI95%         p-unc       BF10     power     abs_r
0   liverind     gonind  pearson   two-sided  563 -0.401865  [-0.47, -0.33]  2.900928e-23  1.269e+20  1.000000  0.401865
1   liverind    ADWBind  pearson   two-sided  563  0.330344     [0.25, 0.4]  8.424595e-16  5.617e+12  1.000000  0.330344
2   liverind    kmethod  pearson   two-sided  563 -0.322689  [-0.39, -0.25]  4.140844e-15  1.175e+12  1.000000  0.322689
3   liverind      place  pearson   two-sided  563 -0.310915  [-0.38, -0.23]  4.393870e-14  1.158e+11  1.000000  0.310915
4   liverind     appind  pearson   two-sided  563  0.302862    [0.23, 0.38]  2.082729e-13  2.519e+10  1.000000  0.302862
5   liverind  gonfatind  pearson   two-sided  563 -0.261939  [-0.34, -0.18]  2.756151e-10  2.243e+07  0.999994  0.261939
6     weight   liverind  pearson   two-sided  563 -0.242173  [-0.32, -0.16]  5.875209e-09  1.146e+06  0.999950  0.242173
7   liverind        BMI  pearson   two-sided  563 -0.235559  [-0.31, -0.16]  1.543712e-08  4.491e+05  0.999903  0.235559
8     gender   liverind  pearson   two-sided  563 -0.212156  [-0.29, -0.13]  3.760299e-07  2.056e+04  0.999162  0.212156
9   liverind    blength  pearson   two-sided  563 -0.185872   [-0.26, -0.1]  9.028133e-06    978.445  0.993684  0.185872
10  liverind   femstate  pearson   two-sided  563  0.155040    [0.07, 0.23]  2.217037e-04     47.438  0.959228  0.155040
11  heartind     sulcer  pearson   two-sided  563  0.127927    [0.05, 0.21]  2.356704e-03      5.316  0.861364  0.127927
12       day   liverind  pearson   two-sided  563 -0.124649  [-0.21, -0.04]  3.050469e-03      4.202  0.843169  0.124649
13  liverind    tailind  pearson   two-sided  563  0.092424    [0.01, 0.17]  2.832090e-02      0.581  0.592966  0.092424
14  liverind     batind  pearson   two-sided  563  0.089661    [0.01, 0.17]  3.341962e-02      0.504  0.567189  0.089661
15  liverind       year  pearson   two-sided  563 -0.074654   [-0.16, 0.01]  7.674379e-02      0.252  0.425258  0.074654
16  heartind    ADWBind  pearson   two-sided  563 -0.061735   [-0.14, 0.02]  1.434798e-01      0.154  0.310206  0.061735
17    gender   heartind  pearson   two-sided  563  0.055793   [-0.03, 0.14]  1.861902e-01      0.126  0.262463  0.055793
18  heartind     appind  pearson   two-sided  563 -0.055045   [-0.14, 0.03]  1.921772e-01      0.123  0.256737  0.055045
19  heartind       year  pearson   two-sided  563 -0.053836   [-0.14, 0.03]  2.021332e-01      0.119  0.247644  0.053836
20  heartind   femstate  pearson   two-sided  563 -0.049505   [-0.13, 0.03]  2.408953e-01      0.105  0.216593  0.049505
21  liverind   heartind  pearson   two-sided  563 -0.040191   [-0.12, 0.04]  3.411480e-01      0.083  0.158577  0.040191
22  liverind     sulcer  pearson   two-sided  563 -0.039984   [-0.12, 0.04]  3.436375e-01      0.083  0.157431  0.039984
23  heartind    tailind  pearson   two-sided  563 -0.039799   [-0.12, 0.04]  3.458781e-01      0.082  0.156410  0.039799
24  heartind     batind  pearson   two-sided  563 -0.039186   [-0.12, 0.04]  3.533691e-01      0.081  0.153062  0.039186
25  heartind        BMI  pearson   two-sided  563 -0.029333   [-0.11, 0.05]  4.872965e-01      0.067  0.106857  0.029333
26  heartind     gonind  pearson   two-sided  563  0.024726   [-0.06, 0.11]  5.582342e-01      0.063  0.090099  0.024726
27  heartind    blength  pearson   two-sided  563  0.020112    [-0.06, 0.1]  6.339376e-01      0.059  0.076345  0.020112
28  heartind    kmethod  pearson   two-sided  563  0.014101    [-0.07, 0.1]  7.384921e-01      0.056  0.062842  0.014101
29  heartind      place  pearson   two-sided  563  0.011192   [-0.07, 0.09]  7.910190e-01      0.055  0.058058  0.011192
30       day   heartind  pearson   two-sided  563 -0.011001   [-0.09, 0.07]  7.945040e-01      0.055  0.057783  0.011001
31    weight   heartind  pearson   two-sided  563 -0.008429   [-0.09, 0.07]  8.418258e-01      0.054  0.054548  0.008429
32  heartind  gonf

* We drop the given rats:

In [38]:
idx=['rat2','rat53','rat120','rat434']
idx=data[data['id'].isin(idx)].index
data=data.drop(idx)
data=data.reset_index(drop=True)

In [39]:
pd.set_option('display.max_rows', 600)

In [40]:
data.head()

id  day  weight  gender  liverind  heartind    appind  femstate  gonfatind    batind  sulcer  kmethod   tailind  blength  place  year   ADWBind    gonind       BMI
0  rat1   67    22.0       1  0.036364  0.005091  0.013636         4   0.000000  0.001864       1        2  0.714286     10.5      1     5  0.454545  0.000000  0.199546
1  rat3  230    37.5       1  0.051200  0.006133  0.020800         4   0.000000  0.001467       2        1  0.904762     10.5      2     8  0.653333  0.000000  0.340136
2  rat4  261    45.0       2  0.047333  0.005556  0.105111        -1   0.293333  0.002178       1        1  0.869565     11.5      4     2  0.260000  2.653242  0.340265
3  rat5  262    43.0       1  0.052558  0.005349  0.021628         4   0.000000  0.001884       1        1  0.720000     12.5      2     8  0.500000  0.000000  0.275200
4  rat6  169    40.0       1  0.057250  0.006500  0.014000         4   0.000000  0.002925       2        1  0.826087     11.5      2     7  0.575000  0.000000  0.302457

* We calculate new correlations

In [41]:
correlation_1=pg.pairwise_corr(data.drop('id',axis=1), method='pearson')


In [42]:
correlation_livheart_1=correlation_given_feautures(correlation_1, livheart_0)
correlation_livheart_1

X          Y   method alternative    n         r           CI95%         p-unc       BF10     power     abs_r
0   heartind     gonind  pearson   two-sided  559 -0.548156   [-0.6, -0.49]  3.680391e-45  6.198e+41  1.000000  0.548156
1   heartind        BMI  pearson   two-sided  559 -0.547770   [-0.6, -0.49]  4.357927e-45  5.241e+41  1.000000  0.547770
2     weight   heartind  pearson   two-sided  559 -0.493509  [-0.55, -0.43]  1.183615e-35  2.311e+32  1.000000  0.493509
3   liverind     gonind  pearson   two-sided  559 -0.476951  [-0.54, -0.41]  4.268225e-33  6.766e+29  1.000000  0.476951
4   heartind      place  pearson   two-sided  559 -0.471492   [-0.53, -0.4]  2.774417e-32   1.06e+29  1.000000  0.471492
5   liverind    ADWBind  pearson   two-sided  559  0.460872    [0.39, 0.52]  9.618731e-31  3.166e+27  1.000000  0.460872
6        day   heartind  pearson   two-sided  559 -0.445350  [-0.51, -0.38]  1.376541e-28  2.329e+25  1.000000  0.445350
7   heartind    blength  pearson   two-sided  559 -0.415817  [-0.48, -0.34]  8.822589e-25  4.009e+21  1.000000  0.415817
8   heartind       year  pearson   two-sided  559 -0.409753  [-0.48, -0.34]  4.806187e-24   7.51e+20  1.000000  0.409753
9   liverind    kmethod  pearson   two-sided  559 -0.396010  [-0.46, -0.32]  1.973438e-22  1.916e+19  1.000000  0.396010
10  liverind      place  pearson   two-sided  559 -0.384740  [-0.45, -0.31]  3.657370e-21  1.075e+18  1.000000  0.384740
11  heartind    kmethod  pearson   two-sided  559 -0.345739  [-0.42, -0.27]  3.861378e-17  1.167e+14  1.000000  0.345739
12  liverind  gonfatind  pearson   two-sided  559 -0.321262  [-0.39, -0.24]  6.916469e-15  7.125e+11  1.000000  0.321262
13  heartind    ADWBind  pearson   two-sided  559  0.309069    [0.23, 0.38]  7.731951e-14  6.673e+10  1.000000  0.309069
14  heartind  gonfatind  pearson   two-sided  559 -0.303140  [-0.38, -0.23]  2.403547e-13  2.196e+10  1.000000  0.303140
15    gender   heartind  pearson   two-sided  559  0.261252    [0.18, 0.34]  3.558516e-10  1.755e+07  0.999993  0.261252
16    gender   liverind  pearson   two-sided  559 -0.247324  [-0.32, -0.17]  3.089648e-09  2.145e+06  0.999968  0.247324
17    weight   liverind  pearson   two-sided  559 -0.244602  [-0.32, -0.17]  4.644007e-09  1.444e+06  0.999958  0.244602
18  liverind    blength  pearson   two-sided  559 -0.237874  [-0.31, -0.16]  1.245034e-08  5.549e+05  0.999916  0.237874
19  heartind   femstate  pearson   two-sided  559 -0.233085  [-0.31, -0.15]  2.467991e-08   2.86e+05  0.999866  0.233085
20  liverind   heartind  pearson   two-sided  559  0.229365    [0.15, 0.31]  4.156930e-08  1.727e+05  0.999808  0.229365
21  heartind     sulcer  pearson   two-sided  559  0.200694    [0.12, 0.28]  1.725439e-06    4774.87  0.997755  0.200694
22  liverind        BMI  pearson   two-sided  559 -0.189724  [-0.27, -0.11]  6.280338e-06   1387.743  0.994949  0.189724
23       day   liverind  pearson   two-sided  559 -0.180948   [-0.26, -0.1]  1.675936e-05    544.892  0.990809  0.180948
24  heartind     batind  pearson   two-sided  559  0.167148    [0.09, 0.25]  7.156101e-05    137.797  0.978409  0.167148
25  liverind   femstate  pearson   two-sided  559  0.161223    [0.08, 0.24]  1.289967e-04     79.115  0.969826  0.161223
26  heartind    tailind  pearson   two-sided  559 -0.124466  [-0.21, -0.04]  3.201965e-03      4.035  0.839543  0.124466
27  liverind       year  pearson   two-sided  559 -0.119792   [-0.2, -0.04]  4.566278e-03      2.925  0.810733  0.119792
28  liverind     appind  pearson   two-sided  559  0.094578    [0.01, 0.18]  2.534367e-02      0.641  0.609730  0.094578
29  liverind    tailind  pearson   two-sided  559  0.083149     [0.0, 0.16]  4.942392e-02      0.363  0.502722  0.083149
30  heartind     appind  pearson   two-sided  559 -0.034424   [-0.12, 0.05]  4.166125e-01      0.074  0.128374  0.034424
31  liverind     sulcer  pearson   two-sided  559 -0.018239    [-0.1, 0.06]  6.669775e-01      0.058  0.071453  0.018239
32  liverind     b

**We see that:**
* before removal most correlation was between liverind and gonind (r= -0.401865) and after removal this correlation didn't change
* after removal most correlation is between heartind and gonind (r =-0.54816) while before removal this correlation was  r = 0.024726	

**Lets calculate the overall changes that has happened:**

In [43]:
correlation_livheart_0.sort_values(['X','Y'],ascending=False,inplace=True)
correlation_livheart_0=correlation_livheart_0.reset_index(drop=True)
correlation_livheart_1.sort_values(['X','Y'],ascending=False,inplace=True)
correlation_livheart_1=correlation_livheart_1.reset_index(drop=True)

correlation_diff=correlation_livheart_0.loc[:,['X','Y','r']]
correlation_diff.loc[:,'r_1']=correlation_livheart_1.loc[:,'r']
correlation_diff.loc[:,'diff']=(correlation_livheart_1.loc[:,'r']-correlation_livheart_0.loc[:,'r'])
correlation_diff.loc[:,'abs_diff']=correlation_diff.loc[:,'diff'].apply(abs)


In [22]:
correlation_diff.sort_values('abs_diff',ascending=False)

X          Y         r       r_1      diff  abs_diff
21  heartind     gonind  0.024726 -0.548156 -0.572881  0.572881
27  heartind        BMI -0.029333 -0.547770 -0.518436  0.518436
1     weight   heartind -0.008429 -0.493509 -0.485080  0.485080
19  heartind      place  0.011192 -0.471492 -0.482684  0.482684
24  heartind    blength  0.020112 -0.415817 -0.435929  0.435929
32       day   heartind -0.011001 -0.445350 -0.434349  0.434349
28  heartind    ADWBind -0.061735  0.309069  0.370804  0.370804
20  heartind    kmethod  0.014101 -0.345739 -0.359840  0.359840
16  heartind       year -0.053836 -0.409753 -0.355916  0.355916
22  heartind  gonfatind  0.007741 -0.303140 -0.310881  0.310881
7   liverind   heartind -0.040191  0.229365  0.269556  0.269556
13  liverind     appind  0.302862  0.094578 -0.208284  0.208284
25  heartind     batind -0.039186  0.167148  0.206334  0.206334
30    gender   heartind  0.055793  0.261252  0.205459  0.205459
23  heartind   femstate -0.049505 -0.233085 -0.183580  0.183580
15  liverind    ADWBind  0.330344  0.460872  0.130528  0.130528
17  heartind    tailind -0.039799 -0.124466 -0.084667  0.084667
12  liverind     batind  0.089661  0.005607 -0.084054  0.084054
8   liverind     gonind -0.401865 -0.476951 -0.075086  0.075086
5   liverind      place -0.310915 -0.384740 -0.073825  0.073825
6   liverind    kmethod -0.322689 -0.396010 -0.073322  0.073322
18  heartind     sulcer  0.127927  0.200694  0.072767  0.072767
9   liverind  gonfatind -0.261939 -0.321262 -0.059324  0.059324
31       day   liverind -0.124649 -0.180948 -0.056300  0.056300
11  liverind    blength -0.185872 -0.237874 -0.052003  0.052003
14  liverind        BMI -0.235559 -0.189724  0.045835  0.045835
2   liverind       year -0.074654 -0.119792 -0.045138  0.045138
29    gender   liverind -0.212156 -0.247324 -0.035169  0.035169
4   liverind     sulcer -0.039984 -0.018239  0.021745  0.021745
26  heartind     appind -0.055045 -0.034424  0.020621  0.020621
3   liverind    tailind  0.092424  0.083149 -0.009274  0.009274
10  liverind   femstate  0.155040  0.161223  0.006183  0.006183
0     weight   liverind -0.242173 -0.244602 -0.002429  0.002429

**We can see that:**
* Most changes happened to correlations involving heartind
* Change is mostly negative meaning that correlations have mostlybecome more negatively oriented

**Lets see the data of those rats**

In [23]:
idx=['rat2','rat53','rat120','rat434']
data_before[data_before['id'].isin(idx)]

id  day  weight  gender  liverind  heartind    appind  femstate  gonfatind    batind  sulcer  kmethod   tailind  blength  place  year   ADWBind    gonind       BMI
1      rat2  251   182.0       1  0.010440  0.004396  0.008791         4   0.023077  0.000742       3        2  0.666667     19.5      3     3  0.302198  1.648659  0.478632
48    rat53   81   104.4       1  0.061303  0.015326  0.009579         4   0.000000  0.002835       2        2  0.666667     15.0      1     5  0.493295  0.000000  0.464000
561  rat120  166   253.0       2  0.035178  3.968379  0.005138        -1   0.013043  0.000794       4        2  0.727273     22.0      3     3  0.189723  1.458615  0.522727
562  rat434  258    55.0       1  0.309273  0.016364  0.075636         4   0.004545  0.003236       1        1  0.829268     20.5      2     8  0.214545  0.223144  0.130874

In [24]:
data.head()

id  day  weight  gender  liverind  heartind    appind  femstate  gonfatind    batind  sulcer  kmethod   tailind  blength  place  year   ADWBind    gonind       BMI
0  rat1   67    22.0       1  0.036364  0.005091  0.013636         4   0.000000  0.001864       1        2  0.714286     10.5      1     5  0.454545  0.000000  0.199546
1  rat3  230    37.5       1  0.051200  0.006133  0.020800         4   0.000000  0.001467       2        1  0.904762     10.5      2     8  0.653333  0.000000  0.340136
2  rat4  261    45.0       2  0.047333  0.005556  0.105111        -1   0.293333  0.002178       1        1  0.869565     11.5      4     2  0.260000  2.653242  0.340265
3  rat5  262    43.0       1  0.052558  0.005349  0.021628         4   0.000000  0.001884       1        1  0.720000     12.5      2     8  0.500000  0.000000  0.275200
4  rat6  169    40.0       1  0.057250  0.006500  0.014000         4   0.000000  0.002925       2        1  0.826087     11.5      2     7  0.575000  0.000000  0.302457

* We can see that heartind of these rats differ much from the other rats. For example rat120 has enormous heartind almost 1000 times more. and the others differ 10-100 times.So this makes correlation seem quite different 
* while liverind for the outlier rats differ too much only for rat2  4-5 times less than the other rats.

### C 
Continue with the data from b), where the listed outliers are removed. Change the special codes for the freezer rats: day=0 and
year=-1. How did the correlations involving either day or year
change? What happens if you remove all freezer rats? What
is your conclusion on the changing correlations involving day or
year?

In [25]:
data_beforeC=data.copy()
data_beforeC.head()

id  day  weight  gender  liverind  heartind    appind  femstate  gonfatind    batind  sulcer  kmethod   tailind  blength  place  year   ADWBind    gonind       BMI
0  rat1   67    22.0       1  0.036364  0.005091  0.013636         4   0.000000  0.001864       1        2  0.714286     10.5      1     5  0.454545  0.000000  0.199546
1  rat3  230    37.5       1  0.051200  0.006133  0.020800         4   0.000000  0.001467       2        1  0.904762     10.5      2     8  0.653333  0.000000  0.340136
2  rat4  261    45.0       2  0.047333  0.005556  0.105111        -1   0.293333  0.002178       1        1  0.869565     11.5      4     2  0.260000  2.653242  0.340265
3  rat5  262    43.0       1  0.052558  0.005349  0.021628         4   0.000000  0.001884       1        1  0.720000     12.5      2     8  0.500000  0.000000  0.275200
4  rat6  169    40.0       1  0.057250  0.006500  0.014000         4   0.000000  0.002925       2        1  0.826087     11.5      2     7  0.575000  0.000000  0.302457

In [26]:
def func (x) :
    if (x==15):
        return -1
    else :
        return x
def func_day (x):
    if(x==400):
        return 0
    else :
        return x

In [27]:
data.loc[:,'year']=data['year'].apply(func)
data.loc[:,'day']=data['day'].apply(func_day)

* We calculate the correlations related to day and year 

In [28]:
correlation_dayyear_0=pg.pairwise_corr(data_beforeC.drop('id',axis=1), method='pearson')
feautures=['day','year']
correlation_dayyear_0=correlation_given_feautures(correlation_dayyear_0, feautures)


In [29]:
correlation_dayyear_1=pg.pairwise_corr(data.drop('id',axis=1), method='pearson')
feautures=['day','year']
correlation_dayyear_1=correlation_given_feautures(correlation_dayyear_1, feautures)


* I create a function to find the difference between correlation coefficients.

In [30]:
def diff_of_correlation(correlation_0,correlation_1):
    correlation_0.sort_values(['X','Y'],ascending=False,inplace=True)
    correlation_0=correlation_0.reset_index(drop=True)
    correlation_1.sort_values(['X','Y'],ascending=False,inplace=True)
    correlation_1=correlation_1.reset_index(drop=True)
    
    correlation_diff=correlation_0.loc[:,['X','Y','r']]
    correlation_diff.loc[:,'r_1']=correlation_1.loc[:,'r']
    corr_diff=(correlation_1.loc[:,'r']-correlation_0.loc[:,'r'])
    correlation_diff.loc[:,'diff']=corr_diff
    
    correlation_diff.loc[:,'abs_diff']=correlation_diff.loc[:,'diff'].apply(abs)
    correlation_diff.sort_values('abs_diff',ascending=False,inplace=True)
    correlation_diff=correlation_diff.reset_index(drop=True)
    return correlation_diff

In [31]:
diff_corr=diff_of_correlation(correlation_dayyear_0,correlation_dayyear_1)

diff_corr

X          Y         r       r_1      diff  abs_diff
0     kmethod       year  0.547255 -0.719957 -1.267212  1.267212
1       place       year  0.464509 -0.725394 -1.189903  1.189903
2         day    kmethod  0.454898 -0.531705 -0.986603  0.986603
3         day      place  0.703645 -0.196415 -0.900060  0.900060
4        year     gonind  0.348625 -0.456100 -0.804725  0.804725
5   gonfatind       year  0.273008 -0.452520 -0.725528  0.725528
6         day     gonind  0.441689 -0.172521 -0.614211  0.614211
7    heartind       year -0.409753  0.193752  0.603505  0.603505
8         day  gonfatind  0.387671 -0.163678 -0.551349  0.551349
9    liverind       year -0.119792  0.388220  0.508012  0.508012
10        day   heartind -0.445350  0.017759  0.463109  0.463109
11       year        BMI  0.431131 -0.003156 -0.434287  0.434287
12   femstate       year  0.232414 -0.189852 -0.422267  0.422267
13       year    ADWBind -0.167799  0.240652  0.408451  0.408451
14        day   liverind -0.180948  0.206799  0.387747  0.387747
15        day        BMI  0.206567 -0.146726 -0.353294  0.353294
16        day   femstate  0.263381 -0.060268 -0.323649  0.323649
17     weight       year  0.236050 -0.085913 -0.321963  0.321963
18        day    ADWBind -0.201892  0.110945  0.312838  0.312838
19     appind       year  0.168379 -0.122115 -0.290494  0.290494
20     batind       year  0.006115 -0.270110 -0.276225  0.276225
21     gender       year -0.113150  0.153056  0.266206  0.266206
22        day     weight  0.118285 -0.139844 -0.258130  0.258130
23        day     batind -0.144641 -0.370528 -0.225887  0.225887
24        day     appind  0.227363  0.007788 -0.219575  0.219575
25    tailind       year  0.292168  0.087634 -0.204533  0.204533
26        day     gender -0.249724 -0.055124  0.194600  0.194600
27        day    tailind  0.278342  0.119594 -0.158748  0.158748
28     sulcer       year -0.203141 -0.053761  0.149380  0.149380
29    blength       year -0.015003 -0.151560 -0.136557  0.136557
30        day       year  0.403535  0.288515 -0.115020  0.115020
31        day     sulcer -0.246864 -0.135272  0.111593  0.111593
32        day    blength -0.003080 -0.107770 -0.104690  0.104690

* We can see that feautures that has changed most are categorized features. since day and year for freezer rats has changed dramatically they affected most to the correlations with the other categorized features. Kmethod and place correlations with day and year are the most changed ones and they all have changed their sign of correlation. Because also year and day for freezer rats has 'changed' sign it seems intuitive.

**lets remove freezer rats** 

In [32]:
data_beforeRemovingFreezer=data.copy()

In [33]:


idx=data[(data['day']==0 ) | (data['year']==-1)].index
data=data.drop(idx)
data=data.reset_index(drop=True)
data.head()


id  day  weight  gender  liverind  heartind    appind  femstate  gonfatind    batind  sulcer  kmethod   tailind  blength  place  year   ADWBind    gonind       BMI
0  rat1   67    22.0       1  0.036364  0.005091  0.013636         4   0.000000  0.001864       1        2  0.714286     10.5      1     5  0.454545  0.000000  0.199546
1  rat3  230    37.5       1  0.051200  0.006133  0.020800         4   0.000000  0.001467       2        1  0.904762     10.5      2     8  0.653333  0.000000  0.340136
2  rat4  261    45.0       2  0.047333  0.005556  0.105111        -1   0.293333  0.002178       1        1  0.869565     11.5      4     2  0.260000  2.653242  0.340265
3  rat5  262    43.0       1  0.052558  0.005349  0.021628         4   0.000000  0.001884       1        1  0.720000     12.5      2     8  0.500000  0.000000  0.275200
4  rat6  169    40.0       1  0.057250  0.006500  0.014000         4   0.000000  0.002925       2        1  0.826087     11.5      2     7  0.575000  0.000000  0.302457

In [34]:
correlation_afterfreezer=pg.pairwise_corr(data.drop('id',axis=1), method='pearson')
feautures=['day','year']
correlation_afterfreezer=correlation_given_feautures(correlation_afterfreezer, feautures)


In [35]:
diff_corr=diff_of_correlation(correlation_dayyear_1,correlation_afterfreezer)

diff_corr.head()

X       Y         r       r_1      diff  abs_diff
0       day   place -0.196415  0.518779  0.715194  0.715194
1  heartind    year  0.193752 -0.137423 -0.331175  0.331175
2       day  gonind -0.172521  0.153537  0.326058  0.326058
3      year  gonind -0.456100 -0.133151  0.322949  0.322949
4       day    year  0.288515 -0.010234 -0.298750  0.298750

* After removal of freezer rats the most and drastic  change has happened in correlation of day and place

In [36]:
data.head()

id  day  weight  gender  liverind  heartind    appind  femstate  gonfatind    batind  sulcer  kmethod   tailind  blength  place  year   ADWBind    gonind       BMI
0  rat1   67    22.0       1  0.036364  0.005091  0.013636         4   0.000000  0.001864       1        2  0.714286     10.5      1     5  0.454545  0.000000  0.199546
1  rat3  230    37.5       1  0.051200  0.006133  0.020800         4   0.000000  0.001467       2        1  0.904762     10.5      2     8  0.653333  0.000000  0.340136
2  rat4  261    45.0       2  0.047333  0.005556  0.105111        -1   0.293333  0.002178       1        1  0.869565     11.5      4     2  0.260000  2.653242  0.340265
3  rat5  262    43.0       1  0.052558  0.005349  0.021628         4   0.000000  0.001884       1        1  0.720000     12.5      2     8  0.500000  0.000000  0.275200
4  rat6  169    40.0       1  0.057250  0.006500  0.014000         4   0.000000  0.002925       2        1  0.826087     11.5      2     7  0.575000  0.000000  0.302457

### E
Continue with the data where all freezer rats are removed. Test
changing special codes of categorical features femstate, kmethod
and place. Can you generate any big changes in correlations involving these features?


In [37]:
data_forchange=data.copy()

* Lets change the femstate from 4 to 0 

In [38]:
def func_2 (x):
    if (x==4):
        return 0
    else :
        return x
data_forchange.loc[:,'femstate']=data_forchange['femstate'].apply(func_2)    
    

In [39]:
correlation_forchange=pg.pairwise_corr(data_forchange.drop('id',axis=1), method='pearson')
feautures=['femstate']
correlation_forchange=correlation_given_feautures(correlation_forchange, feautures)

correlation_forchange_0=pg.pairwise_corr(data.drop('id',axis=1), method='pearson')
feautures=['femstate']
correlation_forchange_0=correlation_given_feautures(correlation_forchange_0, feautures)
diff_corr=diff_of_correlation(correlation_forchange_0,correlation_forchange)

diff_corr

X          Y         r       r_1      diff  abs_diff
0     weight   femstate -0.081535  0.283016  0.364551  0.364551
1   femstate        BMI -0.061782  0.277577  0.339359  0.339359
2   femstate    blength -0.053066  0.258560  0.311627  0.311627
3   femstate     gonind -0.197141  0.045388  0.242530  0.242530
4   heartind   femstate -0.106055 -0.297455 -0.191401  0.191401
5   femstate     batind  0.015534 -0.155038 -0.170572  0.170572
6   femstate    ADWBind  0.322962  0.186908 -0.136055  0.136055
7     gender   femstate -0.889307 -0.768465  0.120842  0.120842
8   femstate    kmethod -0.093995 -0.203342 -0.109347  0.109347
9     appind   femstate  0.146812  0.042936 -0.103876  0.103876
10       day   femstate  0.104815  0.201901  0.097086  0.097086
11  femstate      place  0.065331  0.150173  0.084842  0.084842
12  femstate       year  0.012381  0.096061  0.083679  0.083679
13  liverind   femstate  0.318146  0.388163  0.070017  0.070017
14  femstate    tailind  0.080971  0.018925 -0.062046  0.062046
15  femstate  gonfatind -0.091895 -0.032891  0.059004  0.059004
16  femstate     sulcer -0.027893 -0.064456 -0.036563  0.036563

* As expected correlations with femstate has changed from negative to positive. Because intutively when 4 was for nonpregnant rats their weitgh and femstate was negatively correlated because when femstate is 4 obviously their weight was less than those with femstate 1 (pregnant). when we changed 0 for not pregnant then their weight started to correlate positively with femstate because when femstate was increasing from 0 to 1 then also rats' weight was increasing obviously because of pregnancy.